File: src/main/scala/1.two-sum.scala
object Q1 {
  def twoSum(nums: Array[Int], target: Int): Array[Int] = {
    @scala.annotation.tailrec
    def loop(idx: Int, map: Map[Int, Int]): Array[Int] =
      if (nums.isEmpty || idx >= nums.length) Array.emptyIntArray
      else {
        val h = nums(idx)
        val c = target - h
        if (map.contains(c)) Array(map(c), idx)
        else loop(idx + 1, map + (h -> idx))
      }
    loop(0, Map.empty[Int, Int])
  }
}

File: src/main/java/1.two-sum.java
import java.util.HashMap;
import java.util.Map;

class Q1Java {
  public int[] twoSum(int[] nums, int target) {
    Map<Integer, Integer> map = new HashMap<>();
    for (int i = 0; i < nums.length; i++) {
      int other = target - nums[i];
      if (map.containsKey(other)) {
        return new int[] { map.get(other), i };
      }
      map.put(nums[i], i);
    }
    return null;
  }
}
